src/System.Text.JsonLab/System/Text/Json/JsonReader.cs

```diff
@@ -92,35 +100,97 @@ public bool Read()
             return _isSingleSegment ? ReadSingleSegment(ref _buffer) : ReadMultiSegment();
         }
 
-        private void SkipWhiteSpace()
+        private void SkipWhiteSpaceMultiSegment()
         {
             while (true)
             {
-                byte val = (byte)_reader.Peek();
+                byte val = _buffer[0];
                 if (val != JsonConstants.Space && val != JsonConstants.CarriageReturn && val != JsonConstants.LineFeed && val != JsonConstants.Tab)
                 {
                     break;
                 }
-                _reader.Advance(1);
+                Advance(1);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int ReadVal()
+        {
+            int val = -1;
+            if (_buffer.Length == 0)
+            {
+                if (!TryGetNextSegment())
+                    return val;
             }
+
+            val = _buffer[0];
+            Advance(1);
+            return val;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Advance(int byteCount)
+        {
+            if (byteCount < _buffer.Length)
+            {
+                _buffer = _buffer.Slice(byteCount);
```

Comment:
I'm betting it would be even better if you stored the index instead of slicing every time.

Reply:
I can try that and see if it has significant impact. Slicing the _buffer makes a lot of other operations faster since accesses can start at 0. I wanted to get an order-of-magnitude perf impact of using BufferReader, to see if using it is viable. I think 15% overhead is low enough to continue using it (and benefit from the convenience/less code) and try to improve that instead.